### ISSUE TITLE
NullReferenceException when running an async test for .NET Core

### ISSUE THREAD

**Issue by OleksandrKrutykh** [nunit/nunit#2001]
After upgrading my .NET Core project from NUnit 3.5 to NUnit 3.6, I noticed that a failing `Assert.That` in an async test produces a `NullReferenceException` in the test results instead of a normal assert message. 

The minimal piece of code that reproduces the issue:
```C#
[Test]
public async Task Test()
{
    await Task.Run(() => { });
    Assert.Fail("This error message should be shown");
}
```
Running the test with `dotnet test` command produces the following output:
```
Error : NUnitErrorInvestigation.TestClass.Test
System.NullReferenceException : Object reference not set to an instance of an object.
   at NUnit.Framework.Assert.ReportFailure(String message)
   at NUnitErrorInvestigation.TestClass.<Test>d__0.MoveNext() in C:\Temp\NUnitErrorInvestigation\src\NUnitErrorInvestigation\TestClass.cs:line 13
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at NUnit.Framework.Internal.AsyncInvocationRegion.AsyncTaskInvocationRegion.WaitForPendingOperationsToComplete(Object invocationResult)
   at NUnit.Framework.Internal.Commands.TestMethodCommand.RunAsyncTestMethod(TestExecutionContext context)
```
The project.json file: 
```
{
  "version": "1.0.0-*",
  "testRunner": "nunit",
  "dependencies": {
    "dotnet-test-nunit": "3.4.0-beta-3",
    "Microsoft.NETCore.App": {
      "version": "1.1.0",
      "type": "platform"
    },
    "NUnit": "3.6.0"
  },
  "frameworks": {
    "netcoreapp1.0": {
      "imports": "portable-net45+win8"
    }
  }
}
```
Replacing `"NUnit": "3.6.0"` with `"NUnit": "3.5.0"` in the project.json makes the issue disappear.

The Visual Studio 2015 solution with this failing test: [NUnitErrorInvestigation.zip](https://github.com/nunit/nunit/files/721607/NUnitErrorInvestigation.zip)

I couldn't reproduce the issue if the test was put into a normal "Class Library" project for the .NET Framework 4.6.1. Also, the issue disappears if the `await` is replaced with `Task.Wait` or if `Assert.Fail` is put before the line with `await`.

**Comment by rprouse** [comment]
Thanks for the report. @CharliePoole, this is the second unrelated issue with the same exception and callstack. I think this warrants looking into. I will do so today.

**Comment by OleksandrKrutykh** [comment]
Thank you! The error is similar to one in #1965, but I decided to create a new issue anyway, since the scenario was a bit different.

**Comment by rprouse** [comment]
This is a bit different than #1965, but it does look like a duplicate. It could be that our reworked ReportFailure message is making the issue worse.

**Comment by rprouse** [comment]
Actually, this is so little different than #1965 that I am going to close this as a duplicate. The two issues are cross-linked now, so your information is available for whomever helps to fix #1965. Thanks. 👍
